Dynamo Export use fake tensor

SherlockNoMad opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #94276


This is a prerequisite for dynamo.export() to produce fine graph dynamic shape. 


cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

SherlockNoMad committed:
Dynamo Export use fake tensor

[ghstack-poisoned]

SherlockNoMad committed:
Update on "Dynamo Export use fake tensor"



This is a prerequisite for dynamo.export() to produce fine graph dynamic shape. 


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

SherlockNoMad commented:
> These changes look... plausible, I guess? Is there any test?

Yes, test_dynamic_shapes_right_side in test/dynamo/test_repros.py covers this case.

SherlockNoMad committed:
Update on "Dynamo Export use fake tensor"



This is a prerequisite for dynamo.export() to produce fine graph dynamic shape. 


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented on torch/_dynamo/eval_frame.py:
```diff
@@ -642,22 +646,28 @@ def output(self, target, args, kwargs):
+                graph_with_interpreter,
+                decomposition_table=decomposition_table,
+                _allow_non_fake_inputs=True,
+            )(*retrace_args)
```
I don't really understand this. Deferring to @voznesenskym on the review here

ezyang commented:
I don't feel like the eager side version of fine grained dynamic shapes should depend on this PR. I get the feeling this PR may take a while to review.

voznesenskym commented on torch/_dynamo/eval_frame.py:
```diff
@@ -642,22 +646,28 @@ def output(self, target, args, kwargs):
+        retrace_args = graph_captured_input
+        fake_mode = null_context()
+
+        if tracing_mode == "symbolic":
+            retrace_args = fakified_example_inputs
+            for input in fakified_example_inputs:   # type: ignore[union-attr]
+                if isinstance(input, FakeTensor):
+                    fake_mode = input.fake_mode
+                    break
```
use the util, fake_mode_from_tensors

voznesenskym commented on torch/_dynamo/eval_frame.py:
```diff
@@ -575,9 +577,11 @@ def dynamo_normalization_capturing_compiler(
 
         assert graph is None, "whole graph export entails exactly one graph"
         graph = gm
+        fakified_example_inputs = example_inputs
```
These are compile time inputs, if you want to extract them, please call them that.

SherlockNoMad replied:
renamed to compile_time_inputs

voznesenskym reviewed (changes requested):
This was hard to follow. (1) Please refactor this to indicate that you want `symbolic` to use compile_time args. (2) Is the fake mode logic for symbolic redundant with make_fx?

SherlockNoMad committed:
Update on "Dynamo Export use fake tensor"



This is a prerequisite for dynamo.export() to produce fine graph dynamic shape. 


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

SherlockNoMad committed:
Update on "Dynamo Export use fake tensor"



This is a prerequisite for dynamo.export() to produce fine graph dynamic shape. 


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

SherlockNoMad committed:
Update on "Dynamo Export use fake tensor"



This is a prerequisite for dynamo.export() to produce fine graph dynamic shape. 


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

SherlockNoMad commented:
> This was hard to follow. (1) Please refactor this to indicate that you want `symbolic` to use compile_time args. (2) Is the fake mode logic for symbolic redundant with make_fx?

1) Done. 
2) FakeMode logic for symbolic is not redundant for make_fx, as it's calling make_fx with tracing_mode="real". I tried to fix make_fx, to use FakeTensorMode when inputs is a fake tensor, regardless the tracing_mode spplied, but I couldn't get it to work.

voznesenskym commented on torch/_dynamo/eval_frame.py:
```diff
@@ -631,22 +634,28 @@ def output(self, target, args, kwargs):
+        elif tracing_mode == "symbolic":
+            # For dynamic shape, we need to make_fx through the graph with fake tensors under FakeTensorMode
+            # The fake tensors may contain the fine grain dynamic shape passed down from dynamo
+            fake_mode = fake_mode_from_tensors(compile_time_inputs)
+            with fake_mode:
+                graph = make_fx(
+                    graph_with_interpreter,
+                    decomposition_table=decomposition_table,
+                )(*compile_time_inputs)
```
Can fake mode be a requirement for symbolic in make_fx? Can we push this down?

SherlockNoMad replied:
Notice that, this is calling make_fx with "real" tracing mode using FakeTensor, not "symbolic" tracing mode with FakeTensor. 

I tried pushing FakeMode into make_fx, but I couldn't get it passing the UTs. 
Here's what I have tried: in make_fx, if the tracing model is "real" and if any of the input tensor is a FakeTensor, I set the fakemode.

SherlockNoMad replied:
This is what we do today in aot_autograd. Inside aotautograd, we always ignore user specified tracing mode, and always use the tracing_mode="real". According to @Chillee, it's also expected to use tracing_mode="real" internally. 

See
https://github.com/pytorch/pytorch/blob/master/torch/_functorch/aot_autograd.py#L1053
and
https://github.com/pytorch/pytorch/blob/master/torch/_functorch/aot_autograd.py#L1643

I tried changing the behavior of make_fx, but make_fx is used in so many places, and it's breaks bunch of tests, and it appears that there are subtly of how FakeTensor behave differently under FakeTensorMode vs. non-FakeTensorMode. 

So to me, the easiest path forward is to replicate what AOTAutograd is doing in dynamo.export.

voznesenskym commented on torch/fx/interpreter.py:
```diff
@@ -153,6 +153,7 @@ def run(self, *args, initial_env : Optional[Dict[Node, Any]] = None, enable_io_p
 
     @contextmanager
     def _set_current_node(self, node):
+        self.current_node = node
```
why?

SherlockNoMad replied:
ChangeInputOutputSignature uses self.current_node. 

Alos, having self.current_node available is generally useful.

voznesenskym replied:
Nit, kinda unrelated to this PR. sounds fine otherwise!

voznesenskym commented:
Code looks good! Please just add some more tests, export is a very heavily tested codepath already, but more wont hurt :) very close to a stamp!

voznesenskym commented:
Sherlock assured me 1:1 that existing tests cover this.

SherlockNoMad commented:
@pytorchbot merge

zhxchen17 commented:
@SherlockNoMad have you tested on internal models?

jeanschmidt commented:
@pytorchbot revert -m "break several internal build/test jobs: https://fburl.com/phabricator/1tik7ggb" -c nosignal
